src/pe_source/data/shodan/shodan_search.py

```diff
@@ -0,0 +1,485 @@
+
+    # Insert data into the PE database
+    failed = insert_shodan_data(df, "shodan_assets", thread_name, org_name, failed)
+    failed = insert_shodan_data(
+        risk_df,
+        "shodan_insecure_protocols_unverified_vulns",
+        thread_name,
+        org_name,
+        failed,
+    )
+    failed = insert_shodan_data(
+        vuln_df, "shodan_verified_vulns", thread_name, org_name, failed
+    )
+
+    return failed
+
+
+def is_verified(
+    vulns, cve, av_dict, ac_dict, ci_dict, vuln_data, org_uid, r, d, asn, unverified
+):
+    """Check if a CVE is verified."""
+    v = vulns[cve]
+    if v["verified"]:
+        re = search_circl(cve)
+        r_json = re.json()
+        if r_json is not None:
+            summary = r_json.get("summary", None)
+            product = r_json.get("vulnerable_product", None)
+            attack_vector = r_json.get("access", {}).get("vector")
+            av = av_dict.get(attack_vector, None)
+            attack_complexity = r_json.get("access", {}).get("complexity")
+            ac = ac_dict.get(attack_complexity, None)
+            conf_imp = r_json.get("impact", {}).get("confidentiality")
+            ci = ci_dict.get(conf_imp, None)
+            int_imp = r_json.get("impact", {}).get("integrity")
+            ii = ci_dict.get(int_imp, None)
+            avail_imp = r_json.get("impact", {}).get("availability")
+            ai = ci_dict.get(avail_imp, None)
+            cvss = r_json.get("cvss", None)
+            if cvss == 10:
+                severity = "Critical"
+            elif cvss >= 7:
+                severity = "High"
+            elif cvss >= 4:
+                severity = "Medium"
+            elif cvss > 0:
+                severity = "Low"
+            else:
+                severity = None
+        else:
+            # Set cve info to null if Circl has no results
+            summary = ""
+            product = ""
+            attack_vector = ""
+            av = ""
+            attack_complexity = ""
+            ac = ""
+            conf_imp = ""
+            ci = ""
+            int_imp = ""
+            ii = ""
+            avail_imp = ""
+            ai = ""
+            severity = ""
+            cvss = None
+        vuln_data.append(
+            [
+                org_uid,
+                r["org"],
+                r["ip_str"],
+                d["port"],
+                d["_shodan"]["module"],
+                d["timestamp"],
+                cve,
+                severity,
+                cvss,
+                summary,
+                product,
+                attack_vector,
+                av,
+                attack_complexity,
+                ac,
+                conf_imp,
+                ci,
+                int_imp,
+                ii,
+                avail_imp,
+                ai,
+                r["tags"],
+                r["domains"],
+                r["hostnames"],
+                r["isp"],
+                asn,
+            ]
```

Comment:
I'd definitely recommend using a dictionary here, rather than a large list of values.  Even better, if you use a [`defaultdict`](https://docs.python.org/3/library/collections.html#collections.defaultdict), you can avoid having to initialize each value to `""` and only worry about the one special case (`cvss = None`).

Reply:
https://github.com/cisagov/pe-reports/pull/122/commits/04d36a00f0618fc04d76bbb7365bc54c257b0b58 . Converted to dictionary, but not sure how to use defaultdict here. We want every element in the dict list to include all of the keys so it can easily be converted to a dataframe. Do you have an example?

Reply:
I was mainly talking about being able to clean up this code:
https://github.com/cisagov/pe-reports/blob/04d36a00f0618fc04d76bbb7365bc54c257b0b58/src/pe_source/data/shodan/shodan_search.py#L300-L315

If you create a defaultdict like so:
```
from collections import defaultdict

vuln = defaultdict(lambda: "")
```

Then, if any of the dict keys are accessed before they have been set, they will return an empty string (`""`).

Now your `else:` clause (where Circl has no results) simply becomes:
```
else:
  vuln["cvss"] = None
```
(since all other dict values will default to `""`).

One catch here is that when you assign each value, you would have to tell `get()` to return to empty string if no value is present in the source dict, e.g.:
```
vuln["summary"] = r_json.get("summary", "")
vuln["product"] = r_json.get("vulnerable_product", "")
...
```
If you don't do that, those dict values would get assigned `None`, rather than `""`, which isn't what you want there.

Reply:
Thanks for the example. That makes sense but I'm still a little confused for this particular case:

The problem is we're never directly accessing these keys. We append all the dictionaries to the vuln_data list, then convert that into a df, then insert it into the database. So in this case: 
```
vuln_data = [
{"summary": "summary data", "product": "product data"},
{"summary": "more summary data"}
]
```
There's nothing that'll trigger `"product":""` to be added to the second element because all we run next is pd.DataFrame(vuln_data). product will instead be set to NaN. Unless, we add another step which is counter productive I think. Does that make sense?

Maybe instead of deafultdict we could do something like this:
```
vuln["summary"] = summary or ""  
vuln["product"] = product or ""
```

Reply:
Yes, I understand what you're saying here and I think you're right, so don't worry about the defaultdict stuff here.

However, now that these are all dictionaries instead of tuples, you should definitely alphabetize these long lists of dict keys to improve readability.

Reply:
Sounds good! The dictionaries are now alphabetized here: https://github.com/cisagov/pe-reports/pull/122/commits/ad6ad6c15185bd308297243275175c0ca061b039. Thanks for all your help with this

Reply:
> Sounds good! The dictionaries are now alphabetized here: [ad6ad6c](https://github.com/cisagov/pe-reports/commit/ad6ad6c15185bd308297243275175c0ca061b039). Thanks for all your help with this

FYI, I don't think that is the correct commit.